### PR TITLE
Game class

### PR DESCRIPTION
**Pull request by apple00juice**
Added fields to game class, and created a playerAdapter that stores the player and the players units. As I don't think the player should have a field units because in the lobby he has do units.

### CODE REVIEW

File: src/main/java/ch/uzh/ifi/hase/soprafs22/game/PlayerAdapter.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package ch.uzh.ifi.hase.soprafs22.game;
+
+import ch.uzh.ifi.hase.soprafs22.game.enums.Team;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class PlayerAdapter {
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/game/PlayerAdapter.java:8]: strictly speaking the PlayerAdapter should implement some Interface IPlayer, shouldn't it?

**Review by paolovic — COMMENTED** [review]
You're a beast, do you ever make a pause 😅?
Sorry for let you waiting.

I did some research. I would tend to use a Decorator instead of an Adapter (but maybe I'm "Decorator-biased" and that's why I'm wrong). Why did you choose Adapter over Decorator?

"Adapter changes the interface of an object to adapt it to another interface. Decorator maintains the interface while adding features."
https://stackoverflow.com/questions/17731522/what-are-the-differences-between-the-decorator-and-adapter-patterns

"The Adapter pattern is used for connecting two incompatible interfaces that otherwise cannot be connected directly."
https://www.baeldung.com/java-structural-design-patterns

**Comment by apple00juice** [comment]
> You're a beast, do you ever make a pause 😅? Sorry for let you waiting.
np ^^

I was really struggling with that part, to find a good solution. What I did isn't really the adaptor pattern but I don't know what to name it, maybe more container?

I was also looking at the decorator pattern, but I was not sure if it is possible to add further functionality ( other methods) or if you could just change how a method behaves. They always do the example with the cake or pizza where they overwrite the cost method.

We would then have to work on  running type (the decorator type) to get access to the new functionality, right?

**Comment by paolovic** [comment]
> > You're a beast, do you ever make a pause 😅? Sorry for let you waiting.
> > np ^^
> 
> I was really struggling with that part, to find a good solution. What I did isn't really the adaptor pattern but I don't know what to name it, maybe more container?
> 
> I was also looking at the decorator pattern, but I was not sure if it is possible to add further functionality ( other methods) or if you could just change how a method behaves. They always do the example with the cake or pizza where they overwrite the cost method.
> 
> We would then have to work on running type (the decorator type) to get access to the new functionality, right?

Yep, like the following:
The client (in our case the game) has some components (in our case Interface IPlayer). The class Player (that does not contain the Unit) implements the interface, but at the same time the abstract class BasePlayerDecorator does the same. In addition to that, the BasePlayerDecorator has a property of type IPlayer, so whenever the regular methods (getToken, getName, etc pp) are called, this can be delegated to the player method implementation.
And our concrete PlayerDecorator then extends the BasePlayerDecorator by getUnit and setUnit, so that in a game a Player has some units but not in the lobby.

Bildschirmfoto 2022-04-20 um 11 39 41

**Comment by apple00juice** [comment]
> > > You're a beast, do you ever make a pause 😅? Sorry for let you waiting.
> > > np ^^
> > 
> > 
> > I was really struggling with that part, to find a good solution. What I did isn't really the adaptor pattern but I don't know what to name it, maybe more container?
> > I was also looking at the decorator pattern, but I was not sure if it is possible to add further functionality ( other methods) or if you could just change how a method behaves. They always do the example with the cake or pizza where they overwrite the cost method.
> > We would then have to work on running type (the decorator type) to get access to the new functionality, right?
> 
> Yep, like the following: The client (in our case the game) has some components (in our case Interface IPlayer). The class Player (that does not contain the Unit) implements the interface, but at the same time the abstract class BasePlayerDecorator does the same. In addition to that, the BasePlayerDecorator has a property of type IPlayer, so whenever the regular methods (getToken, getName, etc pp) are called, this can be delegated to the player method implementation. And our concrete PlayerDecorator then extends the BasePlayerDecorator by getUnit and setUnit, so that in a game a Player has some units but not in the lobby.
> 
> Bildschirmfoto 2022-04-20 um 11 39 41
> 
> https://refactoring.guru/design-patterns/decorator

Ahh okey, that makes sense. I think I will change it to that

File: src/main/java/ch/uzh/ifi/hase/soprafs22/game/player/BasePlayerDecorator.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
+package ch.uzh.ifi.hase.soprafs22.game.player;
+
+import ch.uzh.ifi.hase.soprafs22.game.enums.Team;
+import ch.uzh.ifi.hase.soprafs22.user.RegisteredUser;
+
+public class BasePlayerDecorator implements IPlayer {
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/game/player/BasePlayerDecorator.java:6]: abstract 😉

**Review by paolovic — APPROVED** [review]

**Review by hilalcomak — APPROVED** [review]

**Pull request merged**